writeengine/wrapper/writeengine.cpp

```diff
@@ -220,6 +218,178 @@ void WriteEngineWrapper::findSmallestColumn(uint32_t& colId, ColStructList colSt
+                                const void* valArrayVoid, const void* oldValArrayVoid,
+                                BRM::CPInfo* maxMin, bool canStartWithInvalidRange)
+{
+    if (!maxMin)
+    {
+        return ;
+    }
+    bool isUnsigned = false; // TODO: should change with type.
+    switch (colType)
+    {
+        case WR_UBYTE:
+        case WR_USHORT:
+        case WR_UINT:
+        case WR_ULONGLONG:
+        {
+            isUnsigned = true;
+            break;
+        }
+        default:
+            break;
+    }
+    if (!canStartWithInvalidRange)
+    {
+        // check if range is invalid, we can't update it.
+        if (isUnsigned)
+	{
+            if (static_cast<uint64_t>(maxMin->max) < static_cast<uint64_t>(maxMin->min))
+            {
+                return;
+            }
+	}
+	else
+	{
+            if (maxMin->max < maxMin->min)
+            {
+                return;
+            }
+	}
+    }
+    size_t i;
+    for (i = 0; i < totalOldRow; i++) {
+	int64_t value, oldValue;
+        uint64_t uvalue, oldUValue;
+
+	// fetching. May not assign value or oldValue variables when array is not present.
+        switch (colType)
+        {
+            case WR_BYTE:
+            {
+                fetchNewOldValues<int64_t, int8_t>(value, oldValue, valArrayVoid, oldValArrayVoid, i, totalNewRow);
+                break;
+            }
+            case WR_UBYTE:
+            {
+                fetchNewOldValues<uint64_t, uint8_t>(uvalue, oldUValue, valArrayVoid, oldValArrayVoid, i, totalNewRow);
+                break;
+            }
+            case WR_SHORT:
+            {
+                fetchNewOldValues<int64_t, int16_t>(value, oldValue, valArrayVoid, oldValArrayVoid, i, totalNewRow);
+                break;
+            }
+            case WR_USHORT:
+            {
+                fetchNewOldValues<uint64_t, uint16_t>(uvalue, oldUValue, valArrayVoid, oldValArrayVoid, i, totalNewRow);
+                break;
+            }
+            case WR_INT:
+            {
+                fetchNewOldValues<int64_t, int>(value, oldValue, valArrayVoid, oldValArrayVoid, i, totalNewRow);
+                break;
+            }
+            case WR_UINT:
+            {
+                fetchNewOldValues<uint64_t, unsigned int>(uvalue, oldUValue, valArrayVoid, oldValArrayVoid, i, totalNewRow);
+                break;
+            }
+            case WR_LONGLONG:
+            {
+                fetchNewOldValues<int64_t, int64_t>(value, oldValue, valArrayVoid, oldValArrayVoid, i, totalNewRow);
+                break;
+            }
+            case WR_ULONGLONG:
+            {
+                fetchNewOldValues<uint64_t, uint64_t>(uvalue, oldUValue, valArrayVoid, oldValArrayVoid, i, totalNewRow);
+                break;
+            }
+            default:
+                idbassert(0 && "unknown WR type tag");
+                return;
+        }
+        if (isUnsigned)
+        {
+            if (!oldValArrayVoid)
```

Comment:
These if blocks should be moved into a  parametrised function  to reduce boilerplate IMHO. 
Moreover I don't see support for wide-decimal here.

Reply:
I did just that.